events/demolition_derby.py

```diff
@@ -0,0 +1,212 @@
+        """
+        Loads the derby tracking document from disk and initializes game interface type stuff.
+        """
+        super().load_event(doc, spawn_helper, game_interface)
+        doc_text = Path(doc.event_doc_path).read_text()
+        self.event_doc = EventDocument.from_json(doc_text)
+        self.competitors = [Competitor.from_config_path(p) for p in self.event_doc.competitor_cfg_files]
+
+    def save_doc(self):
+        """
+        Writes the derby tracking document to disk.
+        """
+        json = self.event_doc.to_json()
+        path = Path(self.event_meta.event_doc_path)
+        path.write_text(json)
+
+    def init_event(self, competitors: List[Competitor], competition_dir: Path) -> EventMeta:
+        """
+        This should create a document that persists which bots are in the event,
+        and the progress and standings so far. If the program crashes, the
+        doc should be adequate for picking up where we left off.
+
+        Anything using random number generation should be done here.
+        """
+        super().init_event(competitors, competition_dir)
+
+        # arrange all bots in a circle
+        spawn_angles = [2 * math.pi * i / len(competitors) for i in range(len(competitors))]
+        radius = 3000
+        spawn_physics = [Physics(
+            location=Vector3(math.cos(angle) * radius, math.sin(angle) * radius, 50),
+            rotation=Rotator(0, angle + math.pi, 0),
+            velocity=Vector3(0, 0, 0),
+            angular_velocity=Vector3(0, 0, 0)
+        ) for angle in spawn_angles]
+        random.shuffle(spawn_physics)  # randomize spawn positions
+
+        derby_spec = DerbySpecification(perma_death=self.perma_death, max_duration=self.max_duration, starts=spawn_physics)
+        event_doc = EventDocument(
+            derby_spec=derby_spec,
+            competitor_cfg_files=[c.bundle.config_path for c in competitors],
+            result_demolitions={}
+        )
+
+        self.file = self.competition_dir / 'DemolitionDerby.json'
+        self.file.write_text(event_doc.to_json())
+
+        return EventMeta(event_type='DemolitionDerby', event_doc_path=str(self.file))
+
+    def start_derby(self):
+        derby_spec = self.event_doc.derby_spec
+        self.spawn_helper.clear_bots()
+
+        self.on_screen_log.log(f"About to spawn bots for DemolitionDerby.")
+        completed_spawns = self.spawn_helper.spawn_bots([competitor.bundle for competitor in self.competitors])
+
+        self.on_screen_log.log("Waiting for bots to get ready")
+        # We expect a number of ready messages equal to the number of competitors. Doesn't matter in which order they arrive.
+        # Currently we have no way of knowing which bot sent which message, so we don't know who supports this event.
+        for i in range(len(self.competitors)):
+            self.hide_ball()
+            self.spawn_helper.listen_for_events_supported_by_bot(timeout=7)
```

Comment:
If we have 5 AdversityBots who never send ready messages, we'll be waiting for 35 seconds. I might tweak this so we sleep for a few seconds before the loop, and then use a minimal timeout in the listen for events call while looping.